packages/protocol/contracts/stability/GrandaMento.sol

```diff
@@ -180,6 +218,134 @@ contract GrandaMento is
     return proposalId;
   }
 
+  /**
+   * @notice Approves an existing exchange proposal.
+   * @dev Sender must be the approver. Exchange proposal must be in the Proposed state.
+   * @param proposalId The identifier of the proposal to approve.
+   */
+  function approveExchangeProposal(uint256 proposalId) external onlyApprover {
+    ExchangeProposal storage proposal = exchangeProposals[proposalId];
+    // Ensure the proposal is in the Proposed state.
+    require(proposal.state == ExchangeProposalState.Proposed, "Proposal must be in Proposed state");
+    // Set the time the approval occurred and change the state.
+    proposal.approvalTimestamp = block.timestamp;
+    proposal.state = ExchangeProposalState.Approved;
+    emit ExchangeProposalApproved(proposalId);
+  }
+
+  /**
+   * @notice Cancels an exchange proposal.
+   * @dev Only callable by the exchanger if the proposal is in the Proposed state
+   * or the owner if the proposal is in the Approved state.
+   * @param proposalId The identifier of the proposal to cancel.
+   */
+  function cancelExchangeProposal(uint256 proposalId) external nonReentrant {
+    ExchangeProposal storage proposal = exchangeProposals[proposalId];
+    // Require the appropriate state and sender.
+    // This will also revert if a proposalId is given that does not correspond
+    // to a previously created exchange proposal.
+    require(
```

Comment:
Is it easier to debug if we separate this require with a different text for the different cases?

Reply:
Imo it's easily implied by the sender. I like the setup of having a single require statement rather than if statements depending on the state/sender, wdyt?

Reply:
oh you're right, the `||` makes it require more than two requires, I'd just leave like it is 👍